crates/compose/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+
+        let mut arguments = IndexMap::new();
+
+        for (import_name, dependency_info) in dependencies {
+            let (export_name, export_ty) = match dependency_info.export_name {
+                Some(export_name) => {
+                    let Some(export_ty) = self.graph.types()[dependency_info.world_id]
+                        .exports
+                        .get(&export_name)
+                    else {
+                        anyhow::bail!(
+                                "dependency '{dependency_name}' doesn't export '{export_name}' to satisfy import '{import_name}'",
+                                dependency_name = dependency_info.manifest_name,
+                            );
+                    };
+
+                    (export_name, export_ty)
+                }
+                None => {
+                    let Some(export_ty) = self.graph.types()[dependency_info.world_id]
+                        .exports
+                        .get(&import_name)
+                    else {
+                        anyhow::bail!(
+                                "dependency '{dependency_name}' doesn't export '{import_name}' to satisfy import '{import_name}'",
+                                dependency_name = dependency_info.manifest_name,
+                            );
+                    };
+
+                    (import_name.clone(), export_ty)
+                }
+            };
+
+            let import_ty = self.graph.types()[world_id]
+                .imports
+                .get(&import_name)
+                .unwrap();
+
+            // Ensure that export_ty is a subtype of import_ty
+            checker.is_subtype(
+                *export_ty,
+                self.graph.types(),
+                *import_ty,
+                self.graph.types(),
+            ).with_context(|| {
+                format!(
+                    "dependency '{dependency_name}' export '{export_name}' of type '{export_ty:?}' is not a subtype of import '{import_name}' of type '{import_ty:?}'",
+                    dependency_name = dependency_info.manifest_name,
+                )
+            })?;
+
+            let export_id = self
+                .graph
+                .alias_instance_export(dependency_info.instantiation_id, &import_name)?;
+
+            assert!(arguments.insert(import_name, export_id).is_none());
+        }
+
+        Ok(arguments)
+    }
+
+    async fn register_dependency(
+        &mut self,
+        dependency_name: DependencyName,
+        dependency: &LockedComponentDependency,
+    ) -> anyhow::Result<DependencyInfo> {
+        let mut dependency_source = self.loader.load_component(&dependency.source).await?;
+
+        let package_name = dependency_name.package.to_string();
+
+        match dependency.inherit.as_ref() {
+            Some(InheritConfiguration::Some(_configurations)) => {
+                unimplemented!("granular configuration inheritance is not yet supported");
```

Comment:
Nit: shouldn't this be an `anyhow!` error?

Reply:
This code is technically never executed because `Some(InheritConfiguration::Some(_))` is never constructed. It's a variant i anticipate to use soon after landing this to support granular inheritance. If anything i think this should be an `unreachable`.